### PR TITLE
PNG-13

### PR DESCRIPTION
**Pull request by mbuehrig**
Added gulp generator for adding new nodes

### CODE REVIEW

File: config/scaffold/add.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+var gulp = require('gulp');
+var prompt = require('gulp-prompt');
+var tap = require('gulp-tap');
+var path = require('path');
+var yaml = require('js-yaml');
```
Comment (tschuems) [inline config/scaffold/add.js]: remove unused variable

Comment (mbuehrig) [reply config/scaffold/add.js]: Removed...

File: config/scaffold/scaffold.utils.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+var fs = require('fs-extra');
+var path = require('path');
+var yaml = require('js-yaml');
```
Comment (tschuems) [inline config/scaffold/scaffold.utils.js]: remove unused variable

File: config/scaffold/scaffold.utils.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+var _ = require('lodash');
+
+exports.ADDABLE_NODES = ['Component', 'Layout', 'Page']
```
Comment (tschuems) [inline config/scaffold/scaffold.utils.js]: semicolon missing

File: config/scaffold/scaffold.utils.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+exports.getNodeTypePrefix = function(nodeType) {
+  switch (nodeType) {
+    case 'Layout':
+      return 'l_';
+    case 'Page':
+      return 'p_';
+    default:
+      return '';
+  }
+}
```
Comment (tschuems) [inline config/scaffold/scaffold.utils.js]: missing semicolon

File: config/scaffold/add.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      message: 'What would you like to add',
+      choices: scaffoldUtils.ADDABLE_NODES
+    },{
+      type: 'input',
+      name: 'nodeName',
+      message: 'Name?'
+    }], function(res) {
+      nodeName = scaffoldUtils.normalizeNodeName(res.nodeName, res.nodeType);
+      nodeType = res.nodeType;
+    }))
```
Comment (tschuems) [inline config/scaffold/add.js]: missing semicolon

File: config/scaffold/add.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      scaffoldUtils.FIND_REPLACES.forEach(function(replaceItem) {
+        var regex = new RegExp(`{% ${replaceItem.label} %}`, 'g');
+
+        file.contents = Buffer.from(file.contents.toString().replace(regex, nodeName[replaceItem.value]));
+      });
+    }))
+    .pipe(rename({
+      basename: nodeName.fileAndComponent
+    }))
+    .pipe(gulp.dest(`./src/${folder}s/${nodeName.directory}/`))
```
Comment (tschuems) [inline config/scaffold/add.js]: missing semicolon

File: config/scaffold/add.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+})
```
Comment (tschuems) [inline config/scaffold/add.js]: missing semicolon

File: gulpfile.js [hunk 8]
```diff
@@ -3,8 +3,14 @@ var gulp = require('gulp');
 var requireDir = require('require-dir');
 requireDir('./config/gulp');
 
+var add = require('./config/scaffold/add.js');
```
Comment (tschuems) [inline gulpfile.js]: require directory

File: config/scaffold/component/component.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+/**
+ * @name: {% NODE_NAME %}
```
Comment (tschuems) [inline config/scaffold/component/component.js]: use PascalCase for the JS Class

File: config/scaffold/component/component.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+ */
+
+import PinguComponent from '../../assets/scripts/helpers/PinguComponent';
+
+class {% NODE_NAME %} extends PinguComponent {
```
Comment (tschuems) [inline config/scaffold/component/component.js]: use PascalCase for the JS Class

File: config/scaffold/component/component.scss [hunk 11]
```diff
@@ -0,0 +1,8 @@
+/*
+  @name: {{NODE_NAME}}
```
Comment (tschuems) [inline config/scaffold/component/component.scss]: use twig templating language for this

File: config/scaffold/component/component.yml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+config:
+  name: {% NODE_NAME %}
```
Comment (tschuems) [inline config/scaffold/component/component.yml]: use PascalCase for the config.name

File: config/scaffold/add.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      scaffoldUtils.FIND_REPLACES.forEach(function(replaceItem) {
+        var regex = new RegExp(`{% ${replaceItem.label} %}`, 'g');
+
+        file.contents = Buffer.from(file.contents.toString().replace(regex, nodeName[replaceItem.value]));
+      });
+    }))
+    .pipe(rename({
+      basename: nodeName.file
+    }))
+    .pipe(gulp.dest(`./src/${folder}s/${nodeName.directory}/`));
```
Comment (tschuems) [inline config/scaffold/add.js]: @TODO Comment for .src path

File: config/scaffold/add.js [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+var fs = require('fs-extra');
+var rename = require('gulp-rename');
+
+var scaffoldUtils = require('./scaffold.utils.js');
+
+var nodeName = {};
+var nodeType = '';
+
+gulp.task('addGeneratorPrompts', function() {
+  return gulp.src('./config/scaffold/add.js')
```
Comment (tschuems) [inline config/scaffold/add.js]: @TODO comment for config directory

File: config/scaffold/layout/layout.yml [hunk 15]
```diff
@@ -0,0 +1,2 @@
+config:
+  name: {% COMP_NAME %}
```
Comment (tschuems) [inline config/scaffold/layout/layout.yml:2]: do we need the config.name in a layout?

File: config/scaffold/page/page.twig [hunk 16]
```diff
@@ -0,0 +1,5 @@
+{% set layoutPath = '../../layouts/' ~ config.layout|capitalize ~ '/' ~ config.layout ~ '.twig' %}
```
Comment (tschuems) [inline config/scaffold/page/page.twig]: same problem with capitalize -> only makes the first letter Uppercase
but we need PascalCase here aswell

File: gulpfile.js (hunk 8 above)
Comment (mbuehrig) [reply gulpfile.js]: Fixed.

File: config/scaffold/add.js (hunk 13 above)
Comment (mbuehrig) [reply config/scaffold/add.js]: Done

File: config/scaffold/add.js (hunk 14 above)
Comment (mbuehrig) [reply config/scaffold/add.js]: Done

File: config/scaffold/layout/layout.yml (hunk 15 above)
Comment (mbuehrig) [reply config/scaffold/layout/layout.yml:2]: Does it hurt? Yet we don't need it

Comment (tschuems) [reply config/scaffold/layout/layout.yml:2]: ok we leave it for now

**Review by tschuems — DISMISSED** [review]

**Review by tschuems — APPROVED** [review]

**Pull request merged**